pkg/controllers/controlplanemachinesetgenerator/openstack.go

```diff
@@ -0,0 +1,171 @@
+	if err := checkOpenStackMachinesServerGroups(machines); err != nil {
+		return machinev1builder.ControlPlaneMachineSetSpecApplyConfiguration{}, fmt.Errorf("failed to check OpenStack machines ServerGroup: %w", err)
+	}
+
+	controlPlaneMachineSetMachineFailureDomainsApplyConfig, err := buildOpenStackFailureDomains(machineSets, machines)
+	if err != nil {
+		return machinev1builder.ControlPlaneMachineSetSpecApplyConfiguration{}, fmt.Errorf("failed to build ControlPlaneMachineSet's OpenStack failure domains: %w", err)
+	}
+
+	controlPlaneMachineSetMachineSpecApplyConfig, err := buildControlPlaneMachineSetOpenStackMachineSpec(machines)
+	if err != nil {
+		return machinev1builder.ControlPlaneMachineSetSpecApplyConfiguration{}, fmt.Errorf("failed to build ControlPlaneMachineSet's OpenStack spec: %w", err)
+	}
+
+	// We want to work with the newest machine.
+	controlPlaneMachineSetApplyConfigSpec := genericControlPlaneMachineSetSpec(replicas, machines[0].ObjectMeta.Labels[clusterIDLabelKey])
+	controlPlaneMachineSetApplyConfigSpec.Template.OpenShiftMachineV1Beta1Machine.FailureDomains = controlPlaneMachineSetMachineFailureDomainsApplyConfig
+	controlPlaneMachineSetApplyConfigSpec.Template.OpenShiftMachineV1Beta1Machine.Spec = controlPlaneMachineSetMachineSpecApplyConfig
+
+	return controlPlaneMachineSetApplyConfigSpec, nil
+}
+
+// buildOpenStackFailureDomains builds an OpenStackFailureDomain config for the ControlPaneMachineSet from the cluster's Machines and MachineSets.
+func buildOpenStackFailureDomains(machineSets []machinev1beta1.MachineSet, machines []machinev1beta1.Machine) (*machinev1builder.FailureDomainsApplyConfiguration, error) {
+	// Fetch failure domains from the machines
+	machineFailureDomains, err := providerconfig.ExtractFailureDomainsFromMachines(machines)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract failure domains from machines: %w", err)
+	}
+
+	// Fetch failure domains from the machineSets
+	machineSetFailureDomains, err := providerconfig.ExtractFailureDomainsFromMachineSets(machineSets)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract failure domains from machine sets: %w", err)
+	}
+
+	// We have to get rid of duplicates from the failure domains.
+	// We construct a set from the failure domains, since a set can't have duplicates.
+	failureDomains := failuredomain.NewSet(machineFailureDomains...)
+	// Construction of a union of failure domains of machines and machineSets.
+	failureDomains.Insert(machineSetFailureDomains...)
+
+	openStackFailureDomains := []machinev1.OpenStackFailureDomain{}
+	for _, fd := range failureDomains.List() {
+		openStackFailureDomains = append(openStackFailureDomains, fd.OpenStack())
+	}
+
+	emptyFailureDomain := machinev1.OpenStackFailureDomain{}
+
+	// We want to make sure that if a failure domain is empty, it is the only one.
+	if len(openStackFailureDomains) > 1 {
```

Comment:
Might this be cleaner if you used the set to check this? Could drop the inner loop here
```suggestion
	if len(openStackFailureDomains) > 1 && failureDomains.Has(emptyFailureDomain) {
```

Reply:
Ok I'll address this.

Reply:
I have tried several things and couldn't make it work without adding a bunch of more lines... Let me know if I missed something obvious.